[Impl]Ptr-Ref wrapper implementation

curin opened the pull request:
## Table of Contents

* [Code of Conduct](#code-of-conduct)
* [Proposal Implementation](#proposal-implementation)

### Code of Conduct

This project has adopted the code of conduct defined by the [Contributor Covenant](http://contributor-covenant.org/) 
to clarify expected behavior in our community. For more information, see the [.NET Foundation Code of Conduct](http://www.dotnetfoundation.org/code-of-conduct).

## Proposal Implementation

Implemented Ptr, Ptr2D, and Ptr3D and change Mut variants to Ref versions. Ptr does not pin any managed code and is very slimmed down to just wrap native pointers and access their data.

Ptr is used for any variable with the out attribute or is returned, otherwise Ref types are used. All const related code was removed. 

Most of the Ptr tests were removed as Ptr types do not implement conversions to or from managed types. Otherwise, all remaining tests are currently passed by these types.

curin committed:
Ptr-Ref wrapper implementation

-implemented Ptr, Ptr2D, and Ptr3D and change mut variants to Ref versions

Perksey reviewed (commented):
Beautiful, I could basically insta-approve this. Just a couple of things:
- Did we want the Ptr2D/Ptr3D -> string[]/string[][] conversions?
- Can you push the new referencegen

Oh and don't forget about the compilation errors.

curin committed:
Fixed some issues from rebasing

curin committed:
Remove errant comment from previously autogenerated type

curin commented:
So I just realized when I rebased (accidently based this branch on my main which was behind) I forgot to regenerate and I notice some issues I missed when rebasing, should compile all good now.

As for string conversions, we can add those, but I wanted to minimize how much we were allocating in the Ptr types. I would actually more want to remove the string conversion in Ptr/Ptr<T>.

As for referenceGen, none of these are autogenerated. I did miss one of the comments at the top of the ref type. These are all handwritten and modified a bit from the old, generated sources. We could feasibly write a generator for these, but I do feel like it would be more time than it is worth.

curin committed:
Added Array and String Array Conversion functions

-String conversion functions are extensions on allowed types
